Generating keys

mpetrunic opened the pull request:
- random
- from mnemonic and path

GregTheGreek commented:
Thanks for removing the boiler plate 😂

High level looks good, will take a second pass. Only slept 4 hours

wemeetagain commented on src/hdnode.ts:
```diff
@@ -0,0 +1,40 @@
+            return this;
+        }
+        const index = this.getIndex(this.path, this.depth + 1);
+        return new HDNode(deriveChildSK(this.key, index), this.depth + 1, this.path).derive();
```
Something about this seems strange. Can we do this functionally instead of using OOP?

eg:
```typescript
function derive(key: Buffer, path: string): Buffer;

function pathToIndices(path: string): BN[];
```

mpetrunic replied:
not sure what your functions are doing.
Basically HDNode represents node in account tree and derive returns one node deeper (using path and depth to determine what node to derive).

wemeetagain replied:
I don't think it makes sense to create objects in the middle of the path, but rather have an interface that just traverses the path and gives back a key at the end.
```typescript
function derive(key: Buffer, path: string): Buffer {
  let k = key;
  pathToIndices(path).forEach((i) => k = deriveChildSK(k, i));
  return k;
}
``

mpetrunic replied:
Well yeah, that works for eth2 account hierarchy because it's pretty basic but other blockchains might decide to have
different processing on some depths.
Hint, look at eth1:
https://github.com/ethers-io/ethers.js/blob/master/src.ts/utils/hdnode.ts

mpetrunic replied:
not sure whats best way to proceed, others might decided to use similar paths.

GregTheGreek replied:
I think going with this method is actually fine the section in etheres is [here](https://github.com/ethers-io/ethers.js/blob/master/src.ts/utils/hdnode.ts#L185).

wemeetagain replied:
If you look at the [EIP]( https://github.com/ethereum/EIPs/blob/c658eb6cbc0dc7dd7edaa8db429cc9abcf450fa7/EIPS/eip-2334.md
) purpose/motivation sections, it seems like the intent is that different blockchains will use this standard as-is (ie: use the same depth).

tynes replied:
> ```ts
> function derive(key: Buffer, path: string): Buffer;
> 
> function pathToIndices(path: string): BN[];
> ```

Both of these functions make a lot of sense to me. We keep track of the path as an array of u32s. If there is a `'` (or `h`) that follows the integer in the string representation, the first bit of the binary encoding of the u32 gets flipped to a 1.
For `pathToIndices`, you do not need to return BNs since the spec defines u32s which can be represented in JS without loss of precision. See https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

Using `h` over `'` is useful for bash

See https://github.com/bcoin-org/bcoin/blob/85ed59c842f2aa4c1a8154d74a9cad7696cdfee3/lib/hd/common.js#L52

Also:
https://github.com/bcoin-org/bcoin/blob/85ed59c842f2aa4c1a8154d74a9cad7696cdfee3/lib/hd/private.js#L171

wemeetagain commented on src/index.ts:
```diff
@@ -0,0 +1,25 @@
+import {HDNode} from "./hdnode";
+import {Buffer} from "buffer";
+
+export function generateRandomBLSSecretKey(entropy?: Buffer): Buffer {
```
you probably don't need to add "BLS" in the function name here and below

Maybe also worth adding a function comment that explains that `entropy` will be _additional_ entropy, not the entire entropy.

wemeetagain commented on src/index.ts:
```diff
@@ -0,0 +1,25 @@
+    return deriveMasterSK(ikm);
+}
+
+export function fromMnemonicToBlsSecretKey(mnemonic: string, path = "m/12381/60/0/0"): Buffer {
```
I don't think there should be a mnemonicToSeed that doesn't do validation. Ie: You shouldn't get a return value from an invalid mnemonic.

mpetrunic replied:
first line in this function validates mnemonic and you won't get return value

mpetrunic commented on src/key-derivation.ts:
```diff
@@ -2,6 +2,7 @@ import BN from "bn.js";
 import {getHkdf} from "./crypto";
 // @ts-ignore
 import {SHA256} from "bcrypto";
+import {Buffer} from "buffer";
```
this is to ensure whoever is using your library don't need Buffer polyfill. It will make library work out of box. Common practice apparently (take a look at libp2p :) )

wemeetagain replied:
:+1:

tynes commented on package.json:
```diff
@@ -44,8 +49,11 @@
         "bn.js": "^5.0.0",
-        "buffer": "^5.4.3"
+        "buffer": "^5.4.3",
+        "secure-random": "^1.1.2"
```
By the way, there is a `bcrypto` module `bcrypto/lib/random` for generating random numbers. It binds to OpenSSL using `"openssl/rand.h"`
https://github.com/bcoin-org/bcrypto/blob/master/src/random/random.c

It also runs in browsers:
https://github.com/bcoin-org/bcrypto/blob/master/lib/js/random.js

See https://github.com/ChainSafe/bls-ts-key-mgmt/issues/3

mpetrunic commented:
- removed secure-random
- added bcrypto types
- apparently bn.toBuffer doesn't work in browser so it's replaced with Buffer.from(bn.toArray)

Library now works in browser, so if this is ok, I would like to merge it and add readme in next PR

wemeetagain commented:
>    apparently bn.toBuffer doesn't work in browser so it's replaced with Buffer.from(bn.toArray)

there's a `bn.toArrayLike` that i think should be used. maybe get it in another PR

mpetrunic commented:
> > apparently bn.toBuffer doesn't work in browser so it's replaced with Buffer.from(bn.toArray)
> 
> there's a `bn.toArrayLike` that i think should be used. maybe get it in another PR

https://github.com/indutny/bn.js/blob/master/lib/bn.js#L542
same thing

allocate(Buffer) in array like seems to not work
